sourcegraph/sourcegraph.proto

```diff
@@ -288,58 +289,115 @@ message RepoList {
 
 service RepoBadges {
 	// ListBadges lists the available badges for repo.
-	rpc ListBadges(RepoSpec) returns (BadgeList);
+	rpc ListBadges(RepoSpec) returns (BadgeList) {
+		option (google.api.http) = {
+			get: "/v1/repo_badges/list_badges"
+		};
+	};
 
 	// ListCounters lists the available counters for repo.
-	rpc ListCounters(RepoSpec) returns (CounterList);
+	rpc ListCounters(RepoSpec) returns (CounterList) {
+		option (google.api.http) = {
+			get: "/v1/repo_badges/list_counters"
+		};
+	};
 
 	// RecordHit records a visit to a repo (that will be reflected in
 	// its counter).
-	rpc RecordHit(RepoSpec) returns (pbtypes.Void);
+	rpc RecordHit(RepoSpec) returns (pbtypes.Void) {
+		option (google.api.http) = {
+			put: "/v1/repo_badges/record_hit"
+		};
+	};
 
 	// CountHits returns the hit count (optionally in a recent time
 	// period).
-	rpc CountHits(RepoBadgesCountHitsOp) returns (RepoBadgesCountHitsResult);
+	rpc CountHits(RepoBadgesCountHitsOp) returns (RepoBadgesCountHitsResult) {
+		option (google.api.http) = {
+			get: "/v1/repo_badges/count_hits"
+		};
+	};
 }
 
 service RepoStatuses {
 	// GetCombined fetches the combined repository status for the given commit.
-	rpc GetCombined(RepoRevSpec) returns (CombinedStatus);
+	rpc GetCombined(RepoRevSpec) returns (CombinedStatus) {
+		option (google.api.http) = {
+			get: "/v1/repo_statuses/get_combined"
+		};
+	};
 
 	// Create creates a repository status for the given commit.
-	rpc Create(RepoStatusesCreateOp) returns (RepoStatus);
+	rpc Create(RepoStatusesCreateOp) returns (RepoStatus) {
+		option (google.api.http) = {
+			post: "/v1/repo_statuses/post"
+		};
+	};
 }
 
 // Repos exposes information about and actions on both locally hosted
 // and remote repositories.
 service Repos {
 	// Get fetches a repository.
-	rpc Get(RepoSpec) returns (Repo);
+	rpc Get(RepoSpec) returns (Repo) {
+		option (google.api.http) = {
+			get: "/v1/repos/get"
+		};
+	};
 
 	// List repositories.
-	rpc List(RepoListOptions) returns (RepoList);
+	rpc List(RepoListOptions) returns (RepoList) {
+		option (google.api.http) = {
+			get: "/v1/repos/list"
+		};
+	};
 
 	// Create creates a new repository.
-	rpc Create(ReposCreateOp) returns (Repo);
+	rpc Create(ReposCreateOp) returns (Repo) {
+		option (google.api.http) = {
+			post: "/v1/repos/create"
+		};
+	};
 
 	// Delete removes a repository.
-	rpc Delete(RepoSpec) returns (pbtypes.Void);
+	rpc Delete(RepoSpec) returns (pbtypes.Void) {
+		option (google.api.http) = {
+			delete: "/v1/repos/delete"
+		};
+	};
```

Comment:
> Currently, each "get, list, create, delete" operation uses both a different endpoint and a different HTTP verb. Which way is better?

I've looked through the various services just now and it looks like we could do this in a lot of different places throughout the API. It seems to in general be a simplification. so I'll push some commits to switch over to this pattern soon.

It's not obvious to me whether or not whether having the URI/repository-path inside the URL path would be _a benefit_: I imagine that it won't be `GET /v1/repos/www.example.com/user/repo` but rather `GET /v1/repos/www.example.com%2Fuser%2Frepo` because that's what the URL-encoded form would be (and I imagine that grpc-gateway will auto-encode, and inherently decode, all protobuf fields). If true, this means clients will need to URL-encode the path which sounds like an extra step to take over passing it as a query parameter.